### ISSUE TITLE
CheckpointFunctionBackward not called if the output of the checkpointed function is an in-place modified view

### ISSUE THREAD

**Issue by Ivan1248** [pytorch/pytorch#36383]
## 🐛 Bug

When `checkpoint` from `torch.utils.checkpoint` is given a function that returns an in-place modified view tensor (e.g. `x[:].relu_()`), the `grad_fn` attribute of the output of the checkpoint is not set to `CheckpointFunctionBackward` but `AsStridedBackward`.
Consequently, `CheckpointFunctionBackward` is not called and gradient propagation is stopped.

By _view_, I mean an operation that creates a tensor sharing its storage with the input (and assumably uses `as_strided` or something similar internally), e.g. `view`, `squeeze`, `permute`, slicing etc..

Related: #26546, #22072

## To Reproduce

Running the code
``` python
import torch
import torch.utils.checkpoint as cp

param = torch.tensor(1., requires_grad=True)

def f(x):
    print("f called")
    return (x + param)[:].relu()

def g(x):
    print("g called")
    return (x + param).relu_()

def h(x):
    print("h called")
    return (x + param)[:].relu_()  # problematic


for func in [f, g, h]:
    param.grad = None
    x = torch.zeros(5, requires_grad=True)
    y = cp.checkpoint(func, x)
    print(f"grad_fn: {y.grad_fn}")
    y.sum().backward()
    print(f"param.grad: {param.grad}\n")
```
yields the output
```
f called
grad_fn: <torch.autograd.function.CheckpointFunctionBackward object at 0x000001EA03339F28>
f called
param.grad: 5.0

g called
grad_fn: <torch.autograd.function.CheckpointFunctionBackward object at 0x000001EA03339588>
g called
param.grad: 5.0

h called
grad_fn: <AsStridedBackward object at 0x000001EA0351BD88>
param.grad: None
```
`h` is called only once and gradients are not computed.

## Expected behavior

The last iteration of the loop, should be equivalent to the first 2, i.e. `h` should be called twice and gradients should be computed.

## Environment

The output from the [environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py):
```
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Arch Linux
GCC version: (Arch Linux 9.2.1+20200130-2) 9.2.1 20200130
CMake version: version 3.16.5

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration:
GPU 0: GeForce GTX 1070
GPU 1: GeForce GTX 1070

Nvidia driver version: 440.64
cuDNN version: /usr/lib/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.4.0
[pip3] torchvision==0.4.2
```
The bug can also be reproduced with PyTorch 1.4.0 on Windows and PyTorch 1.3.1 on Arch Linux.

## A workaround

Adding another view operation to the output of `h` makes it work as expected. More generally, using the following procedure instead of `cp.checkpoint` can make the code work as desired.

``` python
def robust_checkpoint(function, *args, **kwargs):
    def wrapper(*args_, **kwargs_):
        result = function(*args_, **kwargs_)
        if isinstance(result, torch.Tensor):
            return result[...] if result._version > 0 else result
        else:
            raise NotImplementedError()

    return cp.checkpoint(wrapper, *args, **kwargs)
```

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by Ivan1248** [comment]
#26607 gives more insight about what is happening.

**Comment by albanD** [comment]
This is fixed now. Such inplace operation will properly raise a warning (and soon and error) stating that this will happen.
